src/main/java/org/opensearch/flowframework/util/TenantAwareHelper.java

```diff
@@ -88,4 +94,156 @@ public static String getTenantID(Boolean isMultiTenancyEnabled, RestRequest rest
+     */
+    public static void releaseProvision(String tenantId) {
+        release(tenantId, activeProvisionsPerTenant);
+    }
+
+    /**
+     * Releases a deprovision slot for the given tenant.
+     *
+     * @param tenantId The ID of the tenant for which to release the deprovision slot.
+     */
+    public static void releaseDeprovision(String tenantId) {
+        release(tenantId, activeDeprovisionsPerTenant);
+    }
+
+    /**
+     * Attempts to acquire an execution slot for the given tenant.
+     *
+     * @param tenantId The ID of the tenant requesting the execution slot.
+     * @param executionsMap The map tracking the number of active executions per tenant.
+     * @param maxExecutions The maximum number of simultaneous executions allowed per tenant.
+     * @return true if the execution slot was acquired, false otherwise.
+     */
+    private static boolean tryAcquire(String tenantId, ConcurrentHashMap<String, AtomicInteger> executionsMap, int maxExecutions) {
+        if (tenantId == null) {
+            return true; // No throttling for null tenantId
+        }
+        AtomicInteger count = executionsMap.computeIfAbsent(tenantId, k -> new AtomicInteger(0));
+        if (count.incrementAndGet() <= maxExecutions) {
+            return true;
+        } else {
+            count.decrementAndGet(); // Rollback the increment
+            return false;
+        }
+    }
+
+    /**
+     * Releases an execution slot for the given tenant.
+     *
+     * @param tenantId The ID of the tenant for which to release the execution slot.
+     * @param executionsMap The map tracking the number of active executions per tenant.
+     */
+    private static void release(String tenantId, ConcurrentHashMap<String, AtomicInteger> executionsMap) {
+        if (tenantId == null) {
+            return; // No throttling for null tenantId
+        }
+        AtomicInteger count = executionsMap.get(tenantId);
+        if (count != null) {
+            count.decrementAndGet();
+        }
+    }
+
+    /**
+     * Create an action listener that releases provision throttle on failure only
+     * @param tenantId The tenant ID
+     * @param delegate the wrapped listener
+     * @return a listener wrapping the delegate that calls {@link #releaseProvision(String)} on failure of the wrapped listener.
+     */
+    public static ActionListener<WorkflowResponse> releaseProvisionOnFailureListener(
+        String tenantId,
+        ActionListener<WorkflowResponse> delegate
+    ) {
+        return ActionListener.notifyOnce(new ActionListener<>() {
+            @Override
+            public void onResponse(WorkflowResponse response) {
+                delegate.onResponse(response);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                try {
+                    releaseProvision(tenantId);
+                } finally {
+                    delegate.onFailure(e);
+                }
+            }
+        });
```

Comment:
Same as below

Reply:
We only release on failure here so we can't wrap it with the same method in both cases.